check upload dir permissions

bbodenmiller opened the pull request:
Detect permission and ownership issues with the users uploads dir. This helps troubleshoot and correct [gitlab-org/omnibus-gitlab#311](https://gitlab.com/gitlab-org/omnibus-gitlab/issues/311). Fixes gitlabhq#7500, gitlabhq#7052. Related to gitlabhq#6281. Replaces #8769.

stanhu commented on lib/tasks/gitlab/check.rake:
```diff
@@ -276,6 +277,45 @@ namespace :gitlab do
+        else
+          puts "no".red
+          try_fixing_it(
+            "sudo chown -R #{gitlab_user} #{upload_path}",
```
Does the group name also need to be specified (e.g. `gitlab-www`)? See: https://gitlab.com/gitlab-org/omnibus-gitlab/issues/254

bbodenmiller replied:
Non-Omnibus GitLab upload path does not have a group and Omnibus GitLab fixes the group with `gitlab-ctl reconfigure` per https://github.com/gitlabhq/gitlabhq/pull/8769#discussion_r24588605 so no need to check and fix it here. Also per https://gitlab.com/gitlab-org/omnibus-gitlab/issues/311#note_348792 you wouldn't want to run a recursive applying of the group anyways.

stanhu replied:
Ok, sounds good.

jacobvosmaer replied:
Yeah everything under public/uploads belongs to 'git'. The git:gitlab-www ownership on public/uploads itself is just to give the gitlab-www group read(-only) access.

stanhu commented on lib/tasks/gitlab/check.rake:
```diff
@@ -276,6 +277,45 @@ namespace :gitlab do
+    def check_upload_perms
+      print "Upload directory permissions correct? ... "
+
+      upload_path = File.realpath(Rails.root.join('public/uploads'))
```
`realpath` will throw an exception if the directory does not exist.

bbodenmiller replied:
Suggestions for a fix?

stanhu replied:
You can check that it exists and is a directory first via `File.directory?`.

bbodenmiller replied:
Fixed in latest commit.

bbodenmiller committed:
check uploads dir

Detect issues with uploads dir, e.g. permission and ownership issues
with the users uploads dir. This helps troubleshoot and correct
gitlab-org/omnibus-gitlab#311. Fixes gitlabhq#7500, gitlabhq#7052.
Related to gitlabhq#6281.

stanhu commented:
@jacobvosmaer, can you review this one?

jacobvosmaer commented on lib/tasks/gitlab/check.rake:
```diff
@@ -276,6 +277,57 @@ namespace :gitlab do
+      else
+        puts "no".red
+        try_fixing_it(
+          "sudo chmod 0750 #{upload_path}",
```
Where in the big if-else-if-else are we here? Is the file mode the only thing that's wrong in this code branch, or could the owner also be wrong?

bbodenmiller replied:
Are you proposing changing

`if File.stat(upload_path).mode == 040750`

to

`if File.stat(upload_path).mode == 040750 && File.owned?(upload_path)`?

jacobvosmaer commented:
@bbodenmiller looks nice

@stanhu what is the thing about long argument lists you talk about [above](#commitcomment-12741005)? I don't see it in the code.

stanhu commented:
Never mind; looks like the `-exec command {}` argument sends one command per filename; I thought `find` was bundling the command-line into one long string.

Looks good to me too.

axilleas commented:
@jacobvosmaer @stanhu can we merge this then?

jacobvosmaer commented:
Thanks @bbodenmiller !
